File: core/src/main/java/heast/core/network/NetworkState.java
package heast.core.network;

import heast.core.network.listeners.ClientAuthListener;
import heast.core.network.listeners.ServerAuthListener;
import heast.core.network.listeners.ServerLoginListener;
import heast.core.network.listeners.ClientLoginListener;
import heast.core.network.packets.c2s.auth.*;
import heast.core.network.packets.c2s.login.HelloC2SPacket;
import heast.core.network.packets.c2s.login.KeyC2SPacket;
import heast.core.network.packets.s2c.auth.*;
import heast.core.network.packets.s2c.login.HelloS2CPacket;
import heast.core.network.packets.s2c.login.SuccessS2CPacket;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * A network state defines a state a client can be in the network.
 * States have a unique id and a list of packets that can be sent to the client / server and vice versa.
 */
public enum NetworkState {
    LOGIN(new PacketHandlerInitializer()
        .setup(NetworkSide.CLIENT, new PacketHandler<ServerLoginListener>()
            .register(HelloC2SPacket.class, HelloC2SPacket::new)
            .register(KeyC2SPacket.class, KeyC2SPacket::new)
        )
        .setup(NetworkSide.SERVER, new PacketHandler<ClientLoginListener>()
            .register(HelloS2CPacket.class, HelloS2CPacket::new)
            .register(SuccessS2CPacket.class, SuccessS2CPacket::new)
        )
    ),
    AUTH(new PacketHandlerInitializer()
        .setup(NetworkSide.CLIENT, new PacketHandler<ServerAuthListener>()
            .register(SignupC2SPacket.class, SignupC2SPacket::new)
            .register(LoginC2SPacket.class, LoginC2SPacket::new)
            .register(ResetC2SPacket.class, ResetC2SPacket::new)
            .register(VerifyC2SPacket.class, VerifyC2SPacket::new)
            .register(DeleteC2SPacket.class, DeleteC2SPacket::new)
            .register(LogoutC2SPacket.class, LogoutC2SPacket::new)
        )
        .setup(NetworkSide.SERVER, new PacketHandler<ClientAuthListener>()
            .register(SignupResponseS2CPacket.class, SignupResponseS2CPacket::new)
            .register(LoginResponseS2CPacket.class, LoginResponseS2CPacket::new)
            .register(ResetResponseS2CPacket.class, ResetResponseS2CPacket::new)
            .register(VerifyFailedS2CPacket.class, VerifyFailedS2CPacket::new)
            .register(DeleteResponseS2CPacket.class, DeleteResponseS2CPacket::new)
            .register(LogoutResponseS2CPacket.class, LogoutResponseS2CPacket::new)
        )
    ),
    CHAT(new PacketHandlerInitializer()

    );

    private final Map<NetworkSide, ? extends PacketHandler<? extends PacketListener>> handlers;

    /**
     * Creates a new State with a list of handlers.
     * Handlers are mapped to their respective side in the network, either CLIENT or SERVER.
     * @param initializer The initializer that creates the handlers.
     */
    NetworkState(PacketHandlerInitializer initializer) {
        this.handlers = initializer.handlers;
    }

    /**
     * Gets a handler for a given side.
     * @param side The side to get the handler for.
     * @return The handler for the given side.
     */
    public PacketHandler<? extends PacketListener> getPacketHandler(NetworkSide side) {
        return handlers.get(side);
    }

    public Integer getPacketId(NetworkSide side, Packet<?> packet) {
        return handlers.get(side).getId(packet.getClass());
    }

    /**
     * Utility class for initializing packet handlers.
     * It simply contains a separate handler for each side in the network in form of a map.
     * These can be then registered using the {@link #setup(NetworkSide, PacketHandler)} method.
     */
    public static class PacketHandlerInitializer {
        final Map<NetworkSide, PacketHandler<?>> handlers = new HashMap<>();

        /**
         * Initializes a packet handler for a given side.
         * @param side The side to set up the handler for.
         * @param handler The handler to set up.
         * @return This initializer for chaining.
         * @param <PL> A packet listener extending the {@link PacketListener} interface.
         */
        public <PL extends PacketListener> PacketHandlerInitializer setup(NetworkSide side, PacketHandler<PL> handler) {
            this.handlers.put(side, handler);
            return this;
        }
    }

    /**
     * A packet handler is a function that creates a packet from a packet id and a packet buffer.
     * Packet Handlers can contain multiple packets in the same "area" of the network, and these are stored in the packetIds map.
     * For example, the Login, Signup and Account Reset packets are all handled by the same handler, the {@link ServerLoginListener}.

     * The packetFactories list is used to construct the packets from a {@link PacketBuf} instance.
     * The constructor to use to create the packet is determined by the packetFactory parameter.
     * @see #register(Class, Function)
     *
     * @param <PL> Some packet listener type extending the {@link PacketListener} interface.
     */
    public static class PacketHandler<PL extends PacketListener> {
        private final Map<Class<? extends Packet<PL>>, Integer> packetIds = new HashMap<>();
        private final List<Function<PacketBuf, ? extends Packet<PL>>> packetFactories = new ArrayList<>();

        /**
         * Registers a packet within the handler.
         * @param type The type of packet to register.
         * @param packetFactory The function (constructor) to use to create the packet.
         * @return This handler for chaining.
         * @param <P> Some packet type extending the base Packet.
         */
        public <P extends Packet<PL>> PacketHandler<PL> register(Class<P> type, Function<PacketBuf, P> packetFactory) {
            this.packetIds.put(
                type, packetFactories.size()
            );
            this.packetFactories.add(packetFactory);
            return this;
        }

        /**
         * Gets the packet id for a given packet class.
         * @param packet The packet class to get the according id for.
         * @return The packet id.
         */
        public int getId(Class<?> packet) {
            int id = this.packetIds.get(packet);
            return id < 0 ? -1 : id;
        }

        /**
         * Creates a packet from a packet id and a packet buffer.
         * @param id The packet id.
         * @param buf The packet buffer.
         * @return The packet.
         */
        public Packet<?> createPacket(int id, PacketBuf buf) {
            Function<PacketBuf, ? extends Packet<PL>> fun = this.packetFactories.get(id);
            return fun != null ? fun.apply(buf) : null;
        }
    }
}

File: Core/Network/NetworkState.cs
using Core.Network.Listeners;
using Core.Network.Packets.C2S;
using Core.Network.Packets.S2C;

namespace Core.Network;

public class NetworkState
{
    public static readonly NetworkState Handshake = new NetworkState()
        .Setup(NetworkSide.Server,
            new PacketHandler<IServerHandshakeListener>()
                .Register<HelloC2SPacket>(buf => new HelloC2SPacket(buf))
                .Register<ConnectC2SPacket>(buf => new ConnectC2SPacket(buf))
                .Register<KeyC2SPacket>(buf => new KeyC2SPacket(buf))
                .Register<PingC2SPacket>(buf => new PingC2SPacket(buf)))
        .Setup(NetworkSide.Client,
            new PacketHandler<IClientHandshakeListener>()
                .Register<HelloS2CPacket>(buf => new HelloS2CPacket(buf))
                .Register<ConnectS2CPacket>(buf => new ConnectS2CPacket(buf))
                .Register<KeyS2CPacket>(buf => new KeyS2CPacket(buf))
                .Register<PingS2CPacket>(buf => new PingS2CPacket(buf)));

    public static readonly NetworkState Auth = new NetworkState()
        .Setup(NetworkSide.Server,
            new PacketHandler<IServerAuthListener>()
                .Register<SignupC2SPacket>(buf => new SignupC2SPacket(buf))
                .Register<LoginC2SPacket>(buf => new LoginC2SPacket(buf))
                .Register<ResetC2SPacket>(buf => new ResetC2SPacket(buf))
                .Register<DeleteC2SPacket>(buf => new DeleteC2SPacket(buf))
                .Register<LogoutC2SPacket>(buf => new LogoutC2SPacket(buf))
                .Register<VerifyEmailC2SPacket>(buf => new VerifyEmailC2SPacket(buf))
                .Register<GuestC2SPacket>(buf => new GuestC2SPacket(buf))
                .Register<AccountRequestC2SPacket>(buf => new AccountRequestC2SPacket(buf)))
        .Setup(NetworkSide.Client,
            new PacketHandler<IClientAuthListener>()
                .Register<SignupS2CPacket>(buf => new SignupS2CPacket(buf))
                .Register<VerifyEmailS2CPacket>(buf => new VerifyEmailS2CPacket(buf))
                .Register<AccountRequestS2CPacket>(buf => new AccountRequestS2CPacket(buf)));

    private readonly Dictionary<NetworkSide, IPacketHandler<IPacketListener>> _handlers = new();

    private NetworkState Setup<TPl>(NetworkSide side, IPacketHandler<TPl> handler) where TPl : IPacketListener
    {
        _handlers[side] = (IPacketHandler<IPacketListener>)handler;
        return this;
    }

    public IPacketHandler<IPacketListener> GetPacketHandler(NetworkSide side)
    {
        return _handlers[side];
    }

    public int GetPacketId(NetworkSide side, AbstractPacket packet)
    {
        return GetPacketHandler(side).GetId(packet.GetType());
    }

    public interface IPacketHandler<out TPl> where TPl : IPacketListener
    {
        IPacketHandler<TPl> Register<TP>(Func<PacketBuf, AbstractPacket> packetFactory) where TP : AbstractPacket;
        int GetId(Type packet);
        AbstractPacket? CreatePacket(int id, PacketBuf buf);
    }

    private class PacketHandler<TPl> : IPacketHandler<TPl> where TPl : IPacketListener
    {
        private Dictionary<Type, int> PacketIds { get; } = new();
        private List<Func<PacketBuf, AbstractPacket>> PacketFactories { get; } = new();

        public IPacketHandler<TPl> Register<TP>(Func<PacketBuf, AbstractPacket> packetFactory) where TP : AbstractPacket
        {
            PacketIds[typeof(TP)] = PacketFactories.Count;
            PacketFactories.Add(packetFactory);
            return this;
        }

        public int GetId(Type packet)
        {
            var id = PacketIds[packet];
            return id < 0 ? -1 : id;
        }

        public AbstractPacket? CreatePacket(int id, PacketBuf buf)
        {
            return PacketFactories.Count > id
                ? PacketFactories[id](buf)
                : null;
        }
    }
}